File: long_questions/DSA/Leetcode/golang/4_Sliding_Window/076_Minimum_Window_Substring.go
package main

import "fmt"

// 76. Minimum Window Substring (Variable Size Sliding Window)
// Time: O(N + M), Space: O(1) for ASCII characters
func minWindow(s string, t string) string {
	if len(s) == 0 || len(t) == 0 {
		return ""
	}
	
	// Frequency map for characters in t
	tCount := make(map[byte]int)
	for i := 0; i < len(t); i++ {
		tCount[t[i]]++
	}
	
	// Number of unique characters in t that need to be present in the window
	required := len(tCount)
	
	// Left and right pointers
	left, right := 0, 0
	
	// Keep track of how many unique characters in the current window match the required count
	formed := 0
	
	// Frequency map for the current window
	windowCount := make(map[byte]int)
	
	// Result variables: (window length, left, right)
	result := []int{-1, 0, 0}
	
	for right < len(s) {
		char := s[right]
		windowCount[char]++
		
		// If the frequency of the current character matches exactly the required frequency in t
		if count, exists := tCount[char]; exists && windowCount[char] == count {
			formed++
		}
		
		// Try to contract the window till the point it ceases to be 'desirable'
		for left <= right && formed == required {
			// Save the smallest window
			windowSize := right - left + 1
			if result[0] == -1 || windowSize < result[0] {
				result[0] = windowSize
				result[1] = left
				result[2] = right
			}
			
			// The character at the position left is no longer a part of the window
			leftChar := s[left]
			windowCount[leftChar]--
			if count, exists := tCount[leftChar]; exists && windowCount[leftChar] < count {
				formed--
			}
			
			left++
		}
		
		right++
	}
	
	if result[0] == -1 {
		return ""
	}
	
	return s[result[1] : result[2]+1]
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Variable Size Sliding Window with Frequency Maps
- **Target Frequency Map**: Count required characters from string t
- **Window Frequency Map**: Count characters in current window
- **Formed Counter**: Track how many required characters are satisfied
- **Window Contraction**: Shrink window when all requirements met

## 2. PROBLEM CHARACTERISTICS
- **Substring Containment**: Window must contain all characters from t
- **Frequency Matching**: Must match exact counts of each character
- **Minimum Length**: Find smallest valid window
- **Order Preservation**: Characters must maintain original order in s

## 3. SIMILAR PROBLEMS
- Find All Anagrams in a String (fixed window)
- Longest Substring with At Most K Distinct Characters
- Permutation in String (check if permutation exists)
- Smallest Subsequence of Distinct Characters

## 4. KEY OBSERVATIONS
- **Two-phase approach**: Expand window, then contract when valid
- **Frequency matching**: Need exact character counts, not just presence
- **Optimization opportunity**: Contract as much as possible when valid
- **Early termination**: Can stop if remaining characters insufficient

## 5. VARIATIONS & EXTENSIONS
- **At Most K Characters**: Allow up to K different characters
- **Permutation Check**: Check if any permutation of t exists in s
- **Multiple Queries**: Process multiple t strings efficiently
- **Unicode Support**: Handle international character sets

## 6. INTERVIEW INSIGHTS
- Always clarify: "Character set? Can t have duplicates? Case sensitivity?"
- Edge cases: empty strings, t longer than s, no valid window
- Space complexity: O(1) for ASCII, O(M) for Unicode
- Time complexity: O(N + M) where N=|s|, M=|t|

## 7. COMMON MISTAKES
- Not handling character frequencies correctly (just checking presence)
- Forgetting to update formed counter when shrinking window
- Using nested loops instead of sliding window
- Not handling case where no valid window exists
- Confusing character indices with window boundaries

## 8. OPTIMIZATION STRATEGIES
- **Array for ASCII**: Use fixed-size array instead of hash map
- **Early pruning**: Skip impossible starting positions
- **Character filtering**: Pre-filter s to only relevant characters
- **Optimized contraction**: Shrink window more aggressively

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding the smallest section of a book that contains all required words:**
- You have a long book (string s) and a shopping list (string t)
- You need the smallest continuous section that contains all items from your list
- Start scanning from the beginning, collecting items in your basket
- When you have everything from your list, try to make your basket smaller
- Remove items from the left as long as you still have everything needed
- Keep track of the smallest section that worked

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: String s (source) and string t (target characters)
2. **Goal**: Find smallest window in s containing all characters from t
3. **Output**: The actual substring (empty if none exists)
4. **Constraint**: Must maintain order and frequency of characters

#### Phase 2: Key Insight Recognition
- **"Frequency requirement"** → Need exact counts, not just presence
- **"Two-phase window"** → Expand to meet requirements, then contract
- **"Optimization opportunity"** → Always try to shrink when valid
- **"Character tracking"** → Need to know which requirements are satisfied

#### Phase 3: Strategy Development
```
Human thought process:
"I need to find the smallest window containing all characters from t.
First, I'll count what characters I need from t.
Then I'll expand my window in s until I have all required characters.
Once I have everything, I'll try to shrink the window from the left
while still maintaining all requirements.
I'll keep track of the smallest valid window I find."
```

#### Phase 4: Edge Case Handling
- **Empty strings**: Return empty if either string is empty
- **t longer than s**: Return empty (impossible to contain t)
- **No valid window**: Return empty if t's characters not all in s
- **Single character**: Handle simple case efficiently

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Starting with: s="ADOBECODEBANC", t="ABC"

Human thinking:
"I need A, B, C in my window. Let me scan:

Position 0: 'A' - Have A, need B, C
Position 1: 'D' - Still need B, C
Position 2: 'O' - Still need B, C
Position 3: 'B' - Have A, B, need C
Position 4: 'E' - Still need C
Position 5: 'C' - Now have A, B, C! Window [0,5]="ADOBEC"
           Can I shrink? Remove 'A' (position 0) - lose A, can't shrink
           Smallest so far: "ADOBEC"

Continue expanding:
Position 6: 'O' - Still have A, B, C
Position 7: 'D' - Still have A, B, C
Position 8: 'E' - Still have A, B, C
Position 9: 'B' - Still have A, B, C, try shrinking
           Remove 'A' (position 0) - still have A at position 10? No
           Can't shrink past position 0 yet

Position 10: 'A' - Still have A, B, C, try shrinking
           Remove 'A' (position 0) - still have A at 10, good!
           Window [1,10]="DOBECODEBA", remove 'D' at 1
           Window [2,10]="OBECODEBA", remove 'O' at 2
           Window [3,10]="BECODEBA", remove 'B' at 3 - still have B at 9
           Window [4,10]="ECODEBA", remove 'E' at 4
           Window [5,10]="CODEBA", remove 'C' at 5 - still have C at 5
           Window [6,10]="ODEBA", remove 'O' at 6
           Window [7,10]="DEBA", remove 'D' at 7
           Window [8,10]="EBA", remove 'E' at 8
           Window [9,10]="BA" - lose A, can't shrink further
           New smallest: "CODEBA"

Continue to find "BANC" as final answer...
```

#### Phase 6: Intuition Validation
- **Why sliding window works**: We need contiguous substring
- **Why frequency maps work**: Need exact character counts
- **Why contraction works**: Always try to minimize when valid
- **Why O(N + M)**: Each character visited limited times

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just check presence?"** → Need exact frequencies, not just presence
2. **"Should I shrink after each addition?"** → Only when window is valid
3. **"What about multiple same characters?"** → Frequency maps handle this
4. **"Can I use two pointers only?"** → Need frequency maps for character counts

### Real-World Analogy
**Like finding the smallest section of a recipe book that contains all required ingredients:**
- You have a large recipe book (string s) and a recipe (string t)
- You need the smallest continuous section mentioning all required ingredients
- Start reading from the beginning, highlighting ingredients you find
- When you have all ingredients, try to find a smaller section
- Remove ingredients from the left as long as you still have all required ones
- Keep track of the smallest section that worked

### Human-Readable Pseudocode
```
function minWindow(source, target):
    if source empty or target empty or target longer than source:
        return ""
    
    targetCount = frequency map of target characters
    required = number of unique characters in targetCount
    
    left = 0, right = 0
    formed = 0  // how many required characters are satisfied
    windowCount = empty frequency map
    result = [infinity, 0, 0]  // [length, left, right]
    
    while right < length(source):
        char = source[right]
        windowCount[char]++
        
        // If this character's count now matches required count
        if char in targetCount and windowCount[char] == targetCount[char]:
            formed++
        
        // Try to contract window while it's still valid
        while left <= right and formed == required:
            windowSize = right - left + 1
            if windowSize < result[0]:
                result = [windowSize, left, right]
            
            # Remove character from left
            leftChar = source[left]
            windowCount[leftChar]--
            if leftChar in targetCount and windowCount[leftChar] < targetCount[leftChar]:
                formed--
            left++
        
        right++
    
    if result[0] == infinity:
        return ""
    return source[result[1] : result[2]+1]
```

### Execution Visualization

### Example: s="ADOBECODEBANC", t="ABC"
```
Target counts: A:1, B:1, C:1, required=3

Initial: left=0, right=0, formed=0, windowCount={}

=== EXPANDING PHASE ===
right=0, char='A': windowCount={'A':1}, formed=1
right=1, char='D': windowCount={'A':1,'D':1}, formed=1
right=2, char='O': windowCount={'A':1,'D':1,'O':1}, formed=1
right=3, char='B': windowCount={'A':1,'D':1,'O':1,'B':1}, formed=2
right=4, char='E': windowCount={'A':1,'D':1,'O':1,'B':1,'E':1}, formed=2
right=5, char='C': windowCount={'A':1,'D':1,'O':1,'B':1,'E':1,'C':1}, formed=3

=== CONTRACTION PHASE (formed=3=required) ===
Window [0,5]="ADOBEC", size=6, result=[6,0,5]
Try shrink: remove 'A' at left=0, windowCount['A']=0, formed=2, stop contraction

Continue expanding...
Eventually find "BANC" as optimal solution
```

### Key Visualization Points:
- **Target frequency**: What characters and how many we need
- **Window frequency**: What characters we currently have
- **Formed counter**: How many requirements are satisfied
- **Contraction logic**: Shrink while maintaining validity

### Memory Layout Visualization:
```
Source: A D O B E C O D E B A N C
Index:  0 1 2 3 4 5 6 7 8 9 10 11 12
        ^                 ^
     left=0           right=5
        Window: "ADOBEC" (valid)
```

### Time Complexity Breakdown:
- **Target Processing**: O(M) to build frequency map
- **Window Expansion**: O(N) - each character added once
- **Window Contraction**: O(N) - each character removed at most once
- **Total**: O(N + M) - linear in input sizes
- **Space**: O(K) where K is number of unique characters in t

### Alternative Approaches:

#### 1. Brute Force (O(N²M))
```go
func minWindow(s string, t string) string {
    result := ""
    tCount := make(map[byte]int)
    for _, c := range t {
        tCount[c]++
    }
    
    for i := 0; i < len(s); i++ {
        for j := i; j < len(s); j++ {
            window := s[i:j+1]
            if containsAll(window, tCount) {
                if result == "" || len(window) < len(result) {
                    result = window
                }
                break  // smallest window starting at i
            }
        }
    }
    return result
}
```
- **Pros**: Simple to understand
- **Cons**: O(N²M) time complexity, very slow

#### 2. Optimized with Array (O(N + M) time, O(1) space for ASCII)
```go
func minWindow(s string, t string) string {
    if len(s) == 0 || len(t) == 0 {
        return ""
    }
    
    tCount := [128]int{}
    for _, c := range t {
        tCount[c]++
    }
    
    required := 0
    for _, count := range tCount {
        if count > 0 {
            required++
        }
    }
    
    // Rest of algorithm similar but using arrays
    // ...
}
```
- **Pros**: O(1) space for ASCII, faster operations
- **Cons**: Limited to ASCII character set

### Extensions for Interviews:
- **Permutation in String**: Check if any permutation exists
- **Find All Anagrams**: Return all starting positions
- **At Most K Distinct**: Allow up to K different characters
- **Unicode Support**: Handle international characters
*/
func main() {
	// Test cases
	testCases := []struct {
		s string
		t string
	}{
		{"ADOBECODEBANC", "ABC"},
		{"a", "a"},
		{"a", "aa"},
		{"ab", "b"},
		{"ab", "c"},
		{"abc", "abc"},
		{"aa", "aa"},
		{"ab", "ab"},
		{"bba", "ab"},
		{"aaaaaaaaaaaabbbbbcdd", "abcdd"},
	}
	
	for i, tc := range testCases {
		result := minWindow(tc.s, tc.t)
		fmt.Printf("Test Case %d: s=\"%s\", t=\"%s\" -> Min window: \"%s\"\n", 
			i+1, tc.s, tc.t, result)
	}
}

File: long_questions/DSA/Leetcode/java/4_Sliding_Window/076_Minimum_Window_Substring.java
public class MinimumWindowSubstring {
    
    // 76. Minimum Window Substring
    // Time: O(N + M), Space: O(K) where K is number of unique characters
    public static String minWindow(String s, String t) {
        if (s == null || t == null || s.length() == 0 || t.length() == 0) {
            return "";
        }
        
        // Frequency map for characters in t
        int[] need = new int[128];
        for (char c : t.toCharArray()) {
            need[c]++;
        }
        
        int left = 0, right = 0;
        int minLen = Integer.MAX_VALUE;
        int start = 0;
        int count = t.length();
        
        while (right < s.length()) {
            char c = s.charAt(right);
            
            // If this character is needed
            if (need[c] > 0) {
                count--;
            }
            
            need[c]--;
            right++;
            
            // When we have a valid window
            while (count == 0) {
                // Update minimum window
                if (right - left < minLen) {
                    minLen = right - left;
                    start = left;
                }
                
                // Try to shrink window from left
                char leftChar = s.charAt(left);
                need[leftChar]++;
                if (need[leftChar] > 0) {
                    count++;
                }
                left++;
            }
        }
        
        return minLen == Integer.MAX_VALUE ? "" : s.substring(start, start + minLen);
    }

    public static void main(String[] args) {
        String[][] testCases = {
            {"ADOBECODEBANC", "ABC"},
            {"a", "a"},
            {"a", "aa"},
            {"ab", "b"},
            {"bba", "ab"},
            {"aaaaaaaaaaaabbbbbcddddddeeeee", "abcde"},
            {"", "a"},
            {"a", ""},
            {"abc", "d"},
            {"abacbabc", "abc"},
            {"ab", "ab"},
            {"ab", "ba"},
            {"cabwefgewcwaefgcf", "cae"}
        };
        
        for (int i = 0; i < testCases.length; i++) {
            String s = testCases[i][0];
            String t = testCases[i][1];
            
            String result = minWindow(s, t);
            System.out.printf("Test Case %d: s=\"%s\", t=\"%s\" -> \"%s\"\n", 
                i + 1, s, t, result);
        }
    }
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Sliding Window with Frequency Counting
- **Variable Window**: Window size changes based on validity
- **Two Pointers**: left and right define window boundaries
- **Frequency Tracking**: Character counts for substring validation
- **Minimum Tracking**: Keep track of smallest valid window

## 2. PROBLEM CHARACTERISTICS
- **Substring Problem**: Find contiguous substring containing all target characters
- **Minimum Length**: Among all valid substrings, find shortest
- **Character Frequency**: Need exact counts of each character
- **Optimization**: Expand and shrink window efficiently

## 3. SIMILAR PROBLEMS
- Find All Anagrams in String
- Longest Substring with At Most K Distinct Characters
- Permutation in String
- Longest Repeating Character Replacement

## 4. KEY OBSERVATIONS
- **Window Validity**: When count == 0, current window contains all needed chars
- **Shrinking Strategy**: Try to minimize window while maintaining validity
- **Character Tracking**: Use array for O(1) character access
- **Early Termination**: Can stop when window size exceeds current minimum

## 5. VARIATIONS & EXTENSIONS
- Find all minimum windows (not just one)
- Case-insensitive matching
- Unicode character support
- Multiple pattern matching

## 6. INTERVIEW INSIGHTS
- Clarify: "Should I return empty string if no valid window?"
- Edge cases: empty strings, single character, no common characters
- Time complexity: O(N + M) where N is s.length(), M is t.length()
- Space complexity: O(K) where K is character set size

## 7. COMMON MISTAKES
- Not handling empty input cases
- Incorrect frequency counting (off-by-one errors)
- Not shrinking window properly
- Missing minimum window update

## 8. OPTIMIZATION STRATEGIES
- Use fixed-size array for character frequencies (ASCII assumed)
- Single pass through string with two pointers
- Early termination when window too large
- Minimize string operations for better performance

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding a specific set of letters in a book:**
- You have a book (string s) and need to find specific letters (string t)
- You want the shortest passage containing all those letters
- Use a sliding window to scan through the book
- Track which letters you've found and adjust window size

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: String s (text), String t (pattern to find)
2. **Goal**: Find shortest substring of s containing all characters of t
3. **Output**: Shortest valid substring or empty string

#### Phase 2: Key Insight Recognition
- **"How to track required characters?"** → Frequency array for t
- **"How to know when window is valid?"** → Count of needed characters
- **"How to find minimum?"** → Keep shrinking while valid
- **"How to slide efficiently?"** → Two pointers with frequency updates

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use sliding window with frequency counting:
1. Build frequency map of characters in t
2. Use two pointers (left, right) to define window
3. Expand right pointer to include more characters
4. When window is valid (has all needed chars), try to shrink from left
5. Keep track of minimum window found
6. Continue until right reaches end"
```

#### Phase 4: Algorithm Walkthrough
```
Example: s="ADOBECODEBANC", t="ABC"

Human thinking:
"Build frequency: A:1, B:1, C:1, count=3

Start with left=0, right=0, window=""
Right moves to find characters:
r=0: 'A' (needed) → count=2, need[A]=0
r=1: 'D' (not needed) → need[D]=-1
r=2: 'O' (not needed) → need[O]=-1
r=3: 'B' (needed) → count=1, need[B]=0
r=4: 'E' (not needed) → need[E]=-1
r=5: 'C' (needed) → count=0, need[C]=0

Now count=0, window="ADOBEC" is valid!
Try to shrink from left:
- Remove 'A' (need[A] becomes 1) → count=1, stop shrinking
- Update min window: "ADOBEC" (length 6)

Continue expanding right...
Eventually find "BANC" (length 4) as minimum ✓"
```

### Common Human Pitfalls & How to Avoid Them
1. **"Why not generate all substrings?"** → Too slow O(N³) approach
2. **"What about character encoding?"** → Assume ASCII unless specified
3. **"How to handle duplicates?"** → Frequency counts handle duplicates naturally
4. **"When to stop shrinking?"** → Stop when window becomes invalid

### Real-World Analogy
**Like finding ingredients in a recipe book:**
- You have a recipe book (string s) and need specific ingredients (string t)
- You want the shortest recipe section that mentions all required ingredients
- Scan through the book with a highlighter window
- When you have all ingredients, try to find a shorter section
- Keep track of the shortest complete recipe section found

### Human-Readable Pseudocode
```
function minWindow(s, t):
    if s or t is empty: return ""
    
    need = frequency map of characters in t
    left = 0, right = 0
    count = length of t
    minLen = infinity, start = 0
    
    while right < length of s:
        char = s[right]
        if need[char] > 0:
            count--
        need[char]--
        right++
        
        while count == 0:  // window is valid
            if right - left < minLen:
                minLen = right - left
                start = left
            
            // try to shrink window
            leftChar = s[left]
            if need[leftChar] >= 0:
                count++
            need[leftChar]++
            left++
    
    return s[start:start+minLen] if minLen != infinity else ""
```

### Execution Visualization

### Example: s="ADOBECODEBANC", t="ABC"
```
Window Evolution:
[ADOBEC] - valid (length 6) ✓
[DOBECODEBA] - valid (length 10)
[OBECODEBAN] - valid (length 9)
[BECODEBANC] - valid (length 8)
[ECODEBANC] - valid (length 7)
[CODEBANC] - valid (length 6)
[ODEBANC] - valid (length 5)
[DEBANC] - valid (length 4) ← Minimum!

Final Answer: "BANC"
```

### Key Visualization Points:
- **Frequency Array**: Tracks character requirements in current window
- **Count Variable**: Indicates window validity (0 = valid)
- **Two Pointers**: left and right define current window
- **Minimum Tracking**: Updates when shorter valid window found

### Time Complexity Breakdown:
- **Building Frequency**: O(M) where M is length of t
- **Main Loop**: O(N) where N is length of s
- **Inner Loop**: O(N) total across all iterations (amortized)
- **Total**: O(N + M) time
- **Space**: O(K) where K is character set size (typically 128 for ASCII)
- **Optimal**: Cannot do better than O(N) for this problem
*/
